File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/AbstractViewHolder.java
package com.victorlh.android.framework.listas.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.victorlh.android.framework.listas.ItemLista;

/**
 * @author Victor
 * 17/12/2018
 */
public abstract class AbstractViewHolder<T extends ItemLista> extends RecyclerView.ViewHolder {

	@LayoutRes
	private final int layoutItem;
	@LayoutRes
	private final int layoutItemVoid;

	private ListaAdapter<T> lista;

	private T item;
	private boolean isSeleccionado;

	private ListaAdapter.OnClickListaAdapter onClickListaAdapter;

	public AbstractViewHolder(View itemView, int layoutItem) {
		this(itemView, layoutItem, 0);
	}

	public AbstractViewHolder(View itemView, int layoutItem, int layoutItemVoid) {
		super(itemView);
		this.layoutItem = layoutItem;
		this.layoutItemVoid = layoutItemVoid;
	}

	final void setLista(ListaAdapter<T> listaAdapter) {
		this.lista = listaAdapter;
	}

	public final void procesar(@NonNull T item) {
		this.item = item;

		if (itemView instanceof ViewGroup) {
			Context context = itemView.getContext();
			((ViewGroup) itemView).removeAllViews();

			LayoutInflater inflater = LayoutInflater.from(context);
			View view = inflater.inflate(layoutItem, (ViewGroup) itemView, false);
			ViewGroup.LayoutParams layoutParams = view.getLayoutParams();
			itemView.setLayoutParams(layoutParams);
			((ViewGroup) itemView).addView(view, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));

			if (onClickListaAdapter != null) {
				view.setOnClickListener(onClickListaAdapter);
				view.setOnLongClickListener(onClickListaAdapter);
			}
		}

		onProcesar();
	}

	public final void procesarEmpty() {
		this.item = null;
		if (itemView instanceof ViewGroup) {
			((ViewGroup) itemView).removeAllViews();

			if (layoutItemVoid != 0) {
				Context context = itemView.getContext();
				LayoutInflater inflater = LayoutInflater.from(context);
				View view = inflater.inflate(layoutItemVoid, (ViewGroup) itemView, false);

				ViewGroup.LayoutParams layoutParams = view.getLayoutParams();
				itemView.setLayoutParams(layoutParams);
				((ViewGroup) itemView).addView(view, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
			}
		}
	}

	public T getItem() {
		return item;
	}

	public boolean isSeleccionado() {
		return isSeleccionado;
	}

	protected abstract void onProcesar();

	final void setSeleccion(boolean isSeleccionado) {
		this.isSeleccionado = isSeleccionado;
		if (isSeleccionado) {
			onSelect();
		} else {
			onUnselect();
		}
	}

	public void onSelect() {
	}

	public void onUnselect() {
	}

	protected void notifyItemChange() {
		int itemPosition = lista.getItemPosition(item);
		lista.notifyItemChanged(itemPosition);
	}

	protected ListaAdapter<T> getLista() {
		return lista;
	}

	protected int getPosicion() {
		return lista.getItemPosition(getItem());
	}

	void setOnClickListaAdapter(ListaAdapter.OnClickListaAdapter onClickListaAdapter) {
		this.onClickListaAdapter = onClickListaAdapter;
	}
}

File: listas/src/main/java/com/victorlh/android/framework/listas/adapter/AbstractViewHolder.kt
package com.victorlh.android.framework.listas.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView

/**
 * @author Victor
 * 17/12/2018
 */
abstract class AbstractViewHolder<T> : RecyclerView.ViewHolder {

    private var layoutItem: Int? = null
    private var layoutItemVoid: Int? = null
    private var viewElemento: View? = null
    private var viewVacio: View? = null

    private constructor(itemView: View) : super(itemView)

    constructor(itemView: View, @LayoutRes layoutItem: Int, @LayoutRes layoutItemVoid: Int? = null) : this(itemView) {
        this.layoutItem = layoutItem
        this.layoutItemVoid = layoutItemVoid
    }

    constructor(itemView: View, viewElemento: View, viewVacio: View? = null) : this(itemView) {
        this.viewElemento = viewElemento
        this.viewVacio = viewVacio
    }

    lateinit var lista: AbstractLista<T>

    var item: T? = null
        private set

    var seleccionado = false
        internal set(value) {
            field = value
            if (seleccionado) {
                onSelect()
            } else {
                onUnselect()
            }
        }

    fun procesar(item: T) {
        this.item = item
        if (itemView is ViewGroup) {
            onViewsOldRemove()
            itemView.removeAllViews()

            val view: View = if (viewElemento != null) viewElemento!! else inflateView(layoutItem)

            val layoutParams = view.layoutParams
            itemView.setLayoutParams(layoutParams)
            itemView.addView(view, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
            if (onClickListaAdapter != null) {
                view.setOnClickListener(onClickListaAdapter)
                view.setOnLongClickListener(onClickListaAdapter)
            }
        }
        onProcesar()
    }

    protected fun onViewsOldRemove() {}

    fun procesarEmpty() {
        item = null
        if (itemView is ViewGroup) {
            itemView.removeAllViews()

            val view: View = if (viewVacio != null) viewVacio!! else inflateView(layoutItemVoid)

            val layoutParams = view.layoutParams
            itemView.setLayoutParams(layoutParams)
            itemView.addView(view, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
        }
    }

    protected abstract fun onProcesar()

    open fun onSelect() {}
    open fun onUnselect() {}

    protected fun notifyItemChange() {
        if (posicion >= 0) {
            lista.notifyItemChanged(posicion)
        }
    }

    protected val posicion: Int
        get() = if (item != null) lista.getItemPosition(item!!) else -1

    private var onClickListaAdapter: AbstractLista<T>.OnClickListaAdapter? = null

    internal fun setOnClickListaAdapter(onClickListaAdapter: AbstractLista<T>.OnClickListaAdapter) {
        this.onClickListaAdapter = onClickListaAdapter
    }

    private fun inflateView(layout: Int?): View {
        if (layout == null || itemView !is ViewGroup) {
            return View(itemView.context)
        }
        val inflater = LayoutInflater.from(itemView.context)
        return inflater.inflate(layout, itemView, false)
    }

}